feat(dgw): support TRP streaming

irvingoujAtDevolution opened the pull request:
- **WIP**
- **feat(dgw): support TRP streaming**

CBenoit commented on crates/ascii-streamer/src/trp_decoder.rs:
```diff
@@ -0,0 +1,223 @@
+
+    return Ok((task, AsyncReadChannel::new(rx)));
+
+    fn set_return_type<T, F: Future<Output = T>>(_arg: &F) {}
+    async fn send(
+        sender: &mut tokio::sync::mpsc::Sender<anyhow::Result<String>>,
+        mut json: String,
+    ) -> anyhow::Result<()> {
```
style: I wanted to offer a small suggestion about your coding style. At times, the code can look a bit dense, which makes it harder to scan and understand at a glance. A simple fix is to use two consecutive line breaks between major sections or logical units. E.g., between functions like here, and in the `decode_stream` function defined above.

If you look at the Rust standard library, Tokio, or other major crates, you’ll notice they all do it:

- https://github.com/tokio-rs/tokio/blob/a82bdeebe9560d22a0179ae7ff8ce3986202e24d/tokio/src/io/util/copy_bidirectional.rs#L95
- https://github.com/rust-lang/rust/blob/7a202a9056ea19058a2a4b1b4f508f7f4ec3a0f6/library/alloc/src/vec/mod.rs

Even this message you’re reading now follows the same principle: an extra blank line between paragraphs really helps with readability. This extra spacing groups related statements visually and helps readers quickly grasp the code’s structure.

Would you be open to revisiting this? I believe it will make the code more readable and maintainable overall.

irvingoujAtDevolution replied:
Sorry for my bad practice, I'll look into it.

irvingoujAtDevolution replied:
updated, I'm still grasping how exactly  this style is, let me know if I'm not doing it good enough.

CBenoit commented on devolutions-gateway/src/streaming.rs:
```diff
@@ -83,11 +86,22 @@ pub(crate) async fn stream_file(
             .await
             .with_context(|| format!("failed to open file: {path:?}"))?;
 
+        let path_extension = path.extension().unwrap();
```
suggestion: `path.extension()` is called at many places. Maybe it would be worth moving this at the beginning of the function, with explicit handling of the `None` case.

```rust
let path_extension = path.extension().context("no extension found in the recording file path")?;
```

You’ll save a few `Some(…)` wrapping above as well, and remove an `unwrap()` to audit. (By the way, we have a warning emitted when `unwrap()` is used.)

Also consider mapping `path_extension` to a given `RecordingFileType`, and work with that for the rest of the function. Enums are more ergonomic to work with than strings.

CBenoit reviewed (commented):
thought: Maybe `ascii-streamer` should be `term-streamer` if we’re going to support more protocols? Is ascii appropriate for both asciinema and trp? Thoughts?

irvingoujAtDevolution commented:
> thought: Maybe `ascii-streamer` should be `term-streamer` if we’re going to support more protocols? Is ascii appropriate for both asciinema and trp? Thoughts?

WIth regards to this, the argument could be that the output format is always asciinema v2, so it would not be wired to call this crate `ascii-streamer`, but yeah, `term-streamer` is better.

irvingoujAtDevolution commented on crates/ascii-streamer/examples/decode_trp.rs:
```diff
@@ -0,0 +1,32 @@
+    let output = output
+        .split("=")
+        .last()
+        .ok_or(anyhow::anyhow!("output path is required"))?;
```
updated

irvingoujAtDevolution commented on crates/ascii-streamer/src/lib.rs:
```diff
@@ -69,7 +89,28 @@ pub async fn ascii_stream(
     Ok(())
 }
+
+pub enum Either<A, B> {
```
updated

irvingoujAtDevolution commented on crates/ascii-streamer/src/trp_decoder.rs:
```diff
@@ -0,0 +1,223 @@
+    version: u16,
+    row: u16,
+    col: u16,
+}
```
updated

irvingoujAtDevolution commented on crates/ascii-streamer/src/trp_decoder.rs:
```diff
@@ -0,0 +1,223 @@
+                }
+            })
+            .collect::<String>()
+    }
```
updated

irvingoujAtDevolution commented on crates/ascii-streamer/src/trp_decoder.rs:
```diff
@@ -0,0 +1,223 @@
+                            }
+                            before_setup_cache = None;
+                        } else {
+                            warn!("Termianl setup event cache is empty and we got a setup event");
```
updated

irvingoujAtDevolution commented on crates/ascii-streamer/src/trp_decoder.rs:
```diff
@@ -0,0 +1,223 @@
+        };
+
+        info!("TRP decoder task finished");
+        set_return_type::<anyhow::Result<()>, _>(&task);
```
Yes, it was, but I extracted the async block into a function, this is not needed anymore

irvingoujAtDevolution commented on crates/ascii-streamer/Cargo.toml:
```diff
@@ -5,9 +5,11 @@ edition = "2021"
 anyhow = "1.0"
-serde_json = "1.0"
-tokio = { version = "1.42", features = ["io-util", "sync"] }
+tokio = { version = "1.42", features = ["io-util", "sync", "macros", "rt-multi-thread", "time"] }
```
in `Cargo.toml` or here in Github?

CBenoit replied:
It’s better to document at the relevant places as GitHub PRs are not something we can refer to easily.

irvingoujAtDevolution commented:
Added dependency tokio macro for 
`crates\terminal-streamer\src\lib.rs:64:16`

Added dependency tokio time for 
`crates\terminal-streamer\src\trp_decoder.rs:69:24`

Added dependency tokio rt-multi-thread for 
`crates\terminal-streamer\src\trp_decoder.rs:15:23`

CBenoit reviewed (approved):
Good work, feel free to merge!

CBenoit commented:
> Added dependency tokio macro for `crates\terminal-streamer\src\lib.rs:64:16`
> 
> Added dependency tokio time for `crates\terminal-streamer\src\trp_decoder.rs:69:24`
> 
> Added dependency tokio rt-multi-thread for `crates\terminal-streamer\src\trp_decoder.rs:15:23`

Thank you!

Also see this: https://github.com/Devolutions/devolutions-gateway/pull/1188#discussion_r1915081570
